### PR TITLE
Heap2Local: Properly handle failing array casts

### PR DESCRIPTION
**Pull request by kripken**
Followup to #6727 which added support for failing casts in Struct2Local, but it
turns out that it required Array2Struct changes as well. Specifically, when we
turn an array into a struct then casts can look like they behave differently
(what used to be an array input, becomes a struct), so like with RefTest that we
already handled, check if the cast succeeds in the original form and handle
that.

Found by the fuzzer.

### CODE REVIEW

File: src/passes/Heap2Local.cpp [hunk 1]
```diff
@@ -1050,6 +1062,30 @@ struct Array2Struct : PostWalker<Array2Struct> {
                                         builder.makeConst(Literal(result))));
   }
 
+  void visitRefCast(RefCast* curr) {
```
Comment (tlively) [inline src/passes/Heap2Local.cpp:1065]: Does the comment saying "this is not a problem for ref.cast" above need to be edited?

File: test/lit/passes/heap2local.wast [hunk 2]
```diff
@@ -4303,3 +4303,128 @@
...
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (unreachable)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $array.cast.struct.set
+    (local $eq (ref eq))
+    (local $struct (ref struct))
+    ;; As above, but now the cast fails and is stored to struct. We do not
```
Comment (tlively) [inline test/lit/passes/heap2local.wast:4420]:
```suggestion
    ;; As above, but now the cast fails and is stored to a struct local. We do not
```

**Review by tlively — APPROVED** [review]

File: test/lit/passes/heap2local.wast (hunk 2 above)
Comment (kripken) [reply test/lit/passes/heap2local.wast:4420]: I meant this as "stored to $TYPE", as in "stored to i32" or "stored to anyref"... but maybe that's not grammatical? Anyhow it looks weird, yeah...

File: src/passes/Heap2Local.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/Heap2Local.cpp:1063]: Good catch, fixed.

The cast does go away when we optimize, but now that we allow casts to fail, we do need to know if it fails or not...

File: test/lit/passes/heap2local.wast (hunk 2 above)
Comment (tlively) [reply test/lit/passes/heap2local.wast]: I think it's fine grammatically and would have been clear enough for e.g. i32, but I think the edited version makes it clearer that we are not storing to _a_ struct, which is what I thought was happening when I initially read this.

**Pull request merged**